models/models/output_file.py

```diff
@@ -0,0 +1,216 @@
+from pathlib import Path
+from typing import TypeAlias
+
+from cloudpathlib import AnyPath, CloudPath, GSPath
+from google.cloud.storage import Client
+from pydantic import BaseModel
+
+from models.base import SMBase, parse_sql_bool
+
+RecursiveDict: TypeAlias = dict[str, 'str | RecursiveDict']
+
+
+class OutputFileInternal(SMBase):
+    """File model for internal use"""
+
+    id: int
+    parent_id: int | None = None
+    path: str
+    basename: str
+    dirname: str
+    nameroot: str
+    nameext: str | None
+    file_checksum: str | None
+    size: int
+    meta: dict | None = None
+    valid: bool = False
+    secondary_files: dict | None = {}
+
+    @staticmethod
+    def from_db(**kwargs):
+        """
+        Convert from db keys, mainly converting id to id_
+        """
+        return OutputFileInternal(
+            id=kwargs.pop('id'),
+            parent_id=kwargs.pop('parent_id', None),
+            path=kwargs.get('path'),
+            basename=kwargs.get('basename'),
+            dirname=kwargs.get('dirname'),
+            nameroot=kwargs.get('nameroot'),
+            nameext=kwargs.get('nameext'),
+            file_checksum=kwargs.get('file_checksum'),
+            size=kwargs.get('size'),
+            meta=kwargs.get('meta'),
+            valid=parse_sql_bool(kwargs.get('valid')),
+            secondary_files=kwargs.get('secondary_files', {}),
+        )
+
+    def to_external(self):
+        """
+        Convert to external model
+        """
+        return OutputFile(
+            id=self.id,
+            parent_id=self.parent_id,
+            path=self.path,
+            basename=self.basename,
+            dirname=self.dirname,
+            nameroot=self.nameroot,
+            nameext=self.nameext,
+            file_checksum=self.file_checksum,
+            size=self.size,
+            meta=self.meta,
+            valid=self.valid,
+        )
+
+    @staticmethod
+    async def get_file_info(
```

Comment:
in fact, should this be returning the OutputFile model?

Reply:
It could, but I use the values from the dict to just write to the database, and so it seemed pointless creating an OutputFile object. Is there any benefit from making this change otherwise?

Reply:
I think so, at the moment return type of this function is unknown, so if you were to change the code here to change what keys the dict returns the type checker wouldn't pick up that the code in `create_or_update_output_file` needs to be updated. The best thing about type safety is that you can know the effect area that your code changes have.

Reply:
Makes sense, thanks for explaining!